From 3631fab797c4991e04fb8bf984ca53d00aec520f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jan=20Kundr=C3=A1t?= <jan.kundrat@cesnet.cz>
Date: Tue, 10 Oct 2017 20:49:41 +0200
Subject: [PATCH 2/2] Add another, redundant storage for the environment

---
 include/configs/clearfog.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/configs/clearfog.h b/include/configs/clearfog.h
index eac1e8bbce..8f87eca771 100644
--- a/include/configs/clearfog.h
+++ b/include/configs/clearfog.h
@@ -70,6 +70,9 @@
 #define CONFIG_ENV_OFFSET		0xf0000
 #define CONFIG_ENV_ADDR			CONFIG_ENV_OFFSET
 
+#define CONFIG_ENV_OFFSET_REDUND	0xe0000
+#define CONFIG_SYS_REDUNDAND_ENVIRONMENT
+
 #define CONFIG_PHY_MARVELL		/* there is a marvell phy */
 #define PHY_ANEG_TIMEOUT	8000	/* PHY needs a longer aneg time */
 
-- 
2.14.1

