diff --git a/meson.build b/meson.build
index b338886c38..a968af87ee 100644
--- a/meson.build
+++ b/meson.build
@@ -146,7 +146,7 @@ bootlibdir = join_paths(prefixdir, 'lib/systemd/boot/efi')
 testsdir = join_paths(prefixdir, 'lib/systemd/tests')
 systemdstatedir = join_paths(localstatedir, 'lib/systemd')
 catalogstatedir = join_paths(systemdstatedir, 'catalog')
-randomseeddir = join_paths(localstatedir, 'lib/systemd')
+randomseeddir = '/cfg/random-seed'
 profiledir = join_paths(rootlibexecdir, 'portable', 'profile')
 
 docdir = get_option('docdir')
diff --git a/src/random-seed/random-seed.c b/src/random-seed/random-seed.c
index 0c5f329756..f30f32fb61 100644
--- a/src/random-seed/random-seed.c
+++ b/src/random-seed/random-seed.c
@@ -17,7 +17,7 @@
 #include "string-util.h"
 #include "util.h"
 
-#define POOL_SIZE_MIN 512
+#define POOL_SIZE_MIN 4096
 #define POOL_SIZE_MAX (10*1024*1024)
 
 static int run(int argc, char *argv[]) {
